### PR TITLE
feat: Implement ensureUnknownIsError and extend readme.

### PR DESCRIPTION
**Pull request by yeldiRium**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
 
 You can also optionally let `hydrateResult` hydrate the contained error by passing `potentialErrorConstructors`. This works identically to `hydrateResult`.
 
+## Using the various utilities
+
+### Recognizing errors with `isError`
+
+The function `isError` is used to recognize anything that is derived from the built-in `Error` class. It relies solely on the prototype chain. Use it for example in a catch clause when trying to determine, wether what you have caught is actually an error:
```
Comment (goloroden) [inline README.md:338]:
```suggestion
The function `isError` is used to recognize anything that is derived from the built-in `Error` class. It relies solely on the prototype chain. Use it for example in a `catch` clause when trying to determine, wether what you have caught is actually an error:
```

File: README.md [hunk 2]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+```typescript
+import { isError } from 'defekt';
+
+try {
+    // ...
+} catch (ex: unknown) {
+    if (isError(ex)) {
+        // You can now access ex.message, ex.stack etc.
+    }
+}
```
Comment (goloroden) [inline README.md:349]: The indentation is wrong here.

File: README.md [hunk 3]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+```typescript
+import { isCustomError } from 'defekt';
+
+try {
+    // ...
+} catch (ex: unknown) {
+    if (isCustomError(ex)) {
+        // You can now access ex.message, ex.stack etc but also ex.code.
+    }
+}
```
Comment (goloroden) [inline README.md:367]: Indentation is wrong 😉

File: README.md [hunk 4]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+The function `isError` is used to recognize anything that is derived from the built-in `Error` class. It relies solely on the prototype chain. Use it for example in a catch clause when trying to determine, wether what you have caught is actually an error:
+
+```typescript
+import { isError } from 'defekt';
+
+try {
+    // ...
+} catch (ex: unknown) {
+    if (isError(ex)) {
+        // You can now access ex.message, ex.stack etc.
```
Comment (goloroden) [inline README.md:347]:
```suggestion
        // You can now access ex.message, ex.stack, ...
```

File: README.md [hunk 5]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+You can either identify a general `CustomError`:
+
+```typescript
+import { isCustomError } from 'defekt';
+
+try {
+    // ...
+} catch (ex: unknown) {
+    if (isCustomError(ex)) {
+        // You can now access ex.message, ex.stack etc but also ex.code.
```
Comment (goloroden) [inline README.md:365]:
```suggestion
        // You can now access ex.message, ex.stack, ..., but also ex.code.
```

File: README.md [hunk 6]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+    }
+}
+```
+
+Or you can pass a `CustomError`-constructor to make sure you have a specific type of error in hand:
```
Comment (goloroden) [inline README.md:370]:
```suggestion
Or you can pass a `CustomError` constructor to make sure you have a specific type of error in hand:
```

File: README.md [hunk 7]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+
+```typescript
+import { defekt, isCustomError } from 'defekt';
+
+class MyCustomError extends defekt({ code: 'MyCustomError' }) {}
+
+try {
+    // ...
```
Comment (goloroden) [inline README.md:378]: Indentation 😉

File: README.md [hunk 8]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+} catch (ex: unknown) {
+    if (isCustomError(ex, MyCustomError)) {
+        // In this block ex is of type `MyCustomError`.
+    }
+}
+```
+
+### Making sure something is an error or wrapping it, if not, using `ensureUnknownIsError`
+
+One of the greatest regrets of JavaScript is the ability to throw anything. If you want to bullet-proof your error handling, you need to check that what you catch in a catch-clause is actually an `Error`. `ensureUnknownIsError` takes something you caught and wraps it in an `Error` if necessary. If the caught thing already is an `Error`, `ensureUnknownIsError` returns it unchanged.
```
Comment (goloroden) [inline README.md:388]:
```suggestion
One of the greatest regrets of JavaScript is the ability to throw anything. If you want to bullet-proof your error handling, you need to check that what you catch in a `catch` clause is actually an `Error`. `ensureUnknownIsError` takes something you caught and wraps it in an `Error` if necessary. If the caught thing already is an `Error`, `ensureUnknownIsError` returns it unchanged.
```

File: README.md [hunk 9]
```diff
@@ -331,6 +331,75 @@ if (hydrationResult.hasError()) {
...
+
+```typescript
+import {ensureUnknownIsError} from "./ensureUnknownIsError";
+
+try {
+    // ...
```
Comment (goloroden) [inline README.md:394]: Indentation is wrong 😉

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @yeldiRium 👋 

Besides the change (which is a great idea IMHO), thanks a lot for improving the documentation a lot 😊

However, there are a few suggestions for small changes (and the indentation seems to be broken).

Can you please have a  look at my comments?

**Comment by yeldiRium** [comment]
Ah one of the previous examples had four spaces and I didn't look further than that :D I have corrected the indentation to four spaces everywhere.

**Review by goloroden — APPROVED** [review]
LGTM ✅

**Pull request merged**